### PR TITLE
fix 2d array bug of abiSchemaToJsonSchema in web3-validator

### LINKED ISSUE #6798: 2d array with web3 validator
For below scenario, it will fail in web3 validator
```javascript
    inputs: [
      { internalType: "address[]", name: "addr1", type: "address[]" },
      { internalType: "address[][]", name: "addr2", type: "address[][]" },
    ],
```

```javascript
const addr1 = ['0xaddr']
const addr2 [['0xaddr','0xaddr','0xaddr'],['0xaddr'], ['0xaddr','0xaddr'], ...]
contract.methods.xxMethod(addr1, addr2).estimateGas()
```

check the jsonSchema

```
{"type":"array","items":[{"type":"array","$id":"addr1","items":{"format":"address","required":true}},{"type":"array","items":[{"type":"array","$id":"addr2","items":{"for
mat":"address","required":true}}]}],"maxItems":2,"minItems":2}
```

Also the encodeABI method gave wrong output.

Test in web3.js v1.10.2. Works as expected.

#### Logs

```
code: 1100,
      errors: [
        {
          keyword: 'maxItems',
          instancePath: '/1',
          schemaPath: '#1/maxItems',
          params: { limit: 1 },
          message: 'must NOT have more than 1 items'
        }
      ]
```

#### Environment

web3.js 4.4.0

### PR DESCRIPTION
**Pull request by EtlesL**
## Description
We encountered a bug in the web3-validator code where the handling of multi-dimensional arrays was not implemented correctly.
In the original code snippet, the arraySize vatiable was only assigned the value of the first element in the arraySizes array. This prevented proper handling of multi-dimensional arrays.

```typescript
const arraySize = arraySizes[0];
const item: JsonSchema = {
    type: 'array',
    $id: abiName,
    items: convertEthType(String(baseType)),
    minItems: arraySize,
    maxItems: arraySize,
};
```

To fix the bug, I added an additional loop to iterate over the arraySizes array to handle each dimension:
```typescript
for (let i = 0; i < arraySizes.length; i++) {
  const arraySize = arraySizes[i];
  const item: JsonSchema = {
      type: 'array',
      $id: abiName,
      items: convertEthType(String(baseType)),
      minItems: arraySize,
      maxItems: arraySize,
  };

  if (arraySize < 0) {
      delete item.maxItems;
      delete item.minItems;
  }

  (lastSchema.items as JsonSchema[]).push(item);
}
```

Besides, different from Issue #6486, if the multi-dimensional arrays has a fixed length, it will go to different code snippet.

Fixes #6798 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [x] I ran `npm run lint` with success and extended the tests and types if necessary.
- [x] I ran `npm run test:unit` with success.
- [x] I ran `npm run test:coverage` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [x] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [x] I have updated the `CHANGELOG.md` file in the root folder.
- [x] I have linked Issue(s) with this PR in "Linked Issues" menu.

### CODE REVIEW

File: packages/web3-validator/src/utils.ts [hunk 1]
```diff
@@ -205,37 +205,43 @@ export const abiSchemaToJsonSchema = (
...
-			};
-
-			if (arraySize < 0) {
-				delete item.maxItems;
-				delete item.minItems;
-			}
-
-			(lastSchema.items as JsonSchema[]).push(item);
+		    // iterate over arraySizes array to each dimension
+		    for (let i = 0; i < arraySizes.length; i++) {
```
Comment (jdevcs) [inline packages/web3-validator/src/utils.ts]: @EtlesL 
some of fixes are required for Linter to pass https://github.com/web3/web3.js/actions/runs/8058948881/job/22029093796?pr=6836#step:5:246 , also could you add unit tests for this usecase, Thanks for your contribution,

**Comment by EtlesL** [comment]
@jdevcs 
I'm sorry for the misunderstanding about the type conversion earlier. 
At first, I thought the multi-dimensional array was not converted correctly, so I added a loop. However, after carefully reading the code, I realized that this was not correct. Regarding this bug,
```typescript
public validate(schema: JsonSchema, data: Json, options?: { silent?: boolean }) {
    const zod = convertToZod(schema);
    const result = zod.safeParse(data);
    if (!result.success) {
	    const errors = this.convertErrors(result.error?.issues ?? []);
	    if (errors) {
		    if (options?.silent) {
			    return errors;
		    }
		    throw new Web3ValidatorError(errors);
	    }
    }
    return undefined;
}
```
The core issue lies in the first line, where the schema is converted to Zod. In convertToZod function, multi-dimensional arrays and tuple types are confused.
```typescript
if (schema?.type === 'array' && schema?.items) {
    if (Array.isArray(schema.items) && schema.items.length > 0) {
        // do somsthing
        return z.tuple(arr as [ZodTypeAny, ...ZodTypeAny[]]);
    }
    return z.array(convertToZod(schema.items));
}
```
So, I added additional conditions to distinguish between them. If it's a tuple
1. schema.items MUST be an array
2. the length schema.items MUST have more than 1 items
3. maxItems MUST NOT be undefined
4. the ids of `schema.items` MUST be unique
```typescript
if (Array.isArray(schema.items) 
    && schema.items.length > 1
    && schema.maxItems !== undefined
    && new Set(schema.items.map((item: JsonSchema) => item.$id)).size === schema.items.length) {
        //do something
        return z.tuple(arr as [ZodTypeAny, ...ZodTypeAny[]]);
}
```
Also, in this method, the maxItems and minItems were lost during the conversion, resulting in the length judgment always being "must NOT have more than 1 item."
```typescript
const nextSchema = Array.isArray(schema.items) ? schema.items[0] : schema.items;
let zodArraySchema = z.array(convertToZod(nextSchema));

zodArraySchema = schema.minItems !== undefined ? zodArraySchema.min(schema.minItems) : zodArraySchema;
zodArraySchema = schema.maxItems !== undefined ? zodArraySchema.max(schema.maxItems) : zodArraySchema;
return zodArraySchema;
```

Besides that, the previous unit tests did not cover all test cases. After adding the abiToJsonSchemaCases in the web_validator.test, I found that:
1. The test case "nested tuple object in nested array" has a type error. The tuple[][3] should have 3 elements in the outer layer and an unspecified number in the inner layer, but the original code was written in reverse.
2. The test case "nested tuple object in nested array" also has a small issue with the utils abi to jsonSchema, which I fixed as well.

Please review again, thx.

**Comment by EtlesL** [comment]
I can't run tests of web-eth-contracts on my local environment, I saw the error in the auto pipeline and fixed. Not sure why, because I didn't change the codes of contracts packages. Please review.

**Comment by EtlesL** [comment]
Hi, are there any updates? I noticed the pipeline failed, but there are no error logs provided.

**Comment by jdevcs** [comment]
Thanks for detailed explanation of changes, there are some unstable tests ( not related with this PR ) and we are working in current sprint for fixing these. https://github.com/web3/web3.js/pull/6857, could you rebase your branch when 6857 is merged,

**Comment by EtlesL** [comment]
> Thanks for detailed explanation of changes, there are some unstable tests ( not related with this PR ) and we are working in current sprint for fixing these. #6857, could you rebase your branch when 6857 is merged,

@jdevcs Hi, I saw it merged, and I've updated the branch.

**Review by avkos — APPROVED** [review]

**Review by jdevcs — APPROVED** [review]

**Pull request merged**